cmd/calico-ipam-upgrade/main.go

```diff
@@ -0,0 +1,298 @@
+
+		// Fetch k8s node.
+		k8sNode, err := k8sClient.CoreV1().Nodes().Get(nodename, metav1.GetOptions{})
+		if err != nil {
+			return fmt.Errorf("failed to get k8s node resource: %s", err)
+		}
+
+		// Parse PodCIDR.
+		ip, cidr, err := net.ParseCIDR(k8sNode.Spec.PodCIDR)
+		if err != nil {
+			return fmt.Errorf("PodCIDR %s did not parse successfully: %s", k8sNode.Spec.PodCIDR, err)
+		} else if cidr.Version() == 4 {
+			// We need to get the IP for the podCIDR and increment it to the
+			// first IP in the CIDR.
+			tunIp := ip.To4()
+			if tunIp == nil {
+				return fmt.Errorf("Cannot pick an IPv4 tunnel address from the given CIDR: %s", k8sNode.Spec.PodCIDR)
+			}
+			tunIp[3]++
+			node.Spec.BGP.IPv4IPIPTunnelAddr = tunIp.String()
+
+			// Assign the address via Calico IPAM.
+			ipipTunnelAddr := cnet.ParseIP(node.Spec.BGP.IPv4IPIPTunnelAddr)
+			if err = c.IPAM().AssignIP(ctxt, ipam.AssignIPArgs{
+				IP:       *ipipTunnelAddr,
+				Hostname: nodename,
+				Attrs: map[string]string{
+					"node":           nodename,
+					"ipipTunnelAddr": "true",
+				},
+			}); err != nil {
+				if _, ok := err.(errors.ErrorResourceAlreadyExists); !ok {
+					return fmt.Errorf("failed to get add IPIP tunnel addr %s: %s", node.Spec.BGP.IPv4IPIPTunnelAddr, err)
+				}
+				log.Info("IPIP tunnel address already assigned, continuing...")
+			}
+
+			// Save the calico node object to the datastore.
+			if node, err = c.Nodes().Update(ctxt, node, options.SetOptions{}); err != nil {
+				return fmt.Errorf("failed to save newly updated node object: %s", err)
+			}
+
+			log.Info("Successfully assigned and updated IPIP tunnel address")
+		}
+	}
+
+	// Check to see if the system is still using host-local
+	// by checking the existence of the path.
+	if _, err := os.Stat(ipAllocPath); err != nil && os.IsNotExist(err) {
+		log.Info("k8s-pod-network dir not found; no migration necessary, successfully exiting...")
+		return nil
+	}
+
+	// Check if the directory is empty.
+	ipamDir, err := os.Open(ipAllocPath)
+	if err != nil {
+		return fmt.Errorf("failed to open k8s-pod-network dir: %s", err)
+	}
+	if _, err = ipamDir.Readdirnames(1); err != nil {
+		if err == io.EOF {
+			log.Info("k8s-pod-network dir empty; no migration necessary, successfully exiting...")
+			return nil
+		}
+		return fmt.Errorf("failed to read k8s-pod-network dir names: %s", err)
+	}
+	defer func() {
+		if err = ipamDir.Close(); err != nil {
+			log.WithError(err).Errorf("failed to close k8s-pod-network directory")
+		}
+	}()
+
+	// Disable cni by setting datastoreReady to false.
+	clusterInfo, err := c.ClusterInformation().Get(ctxt, "default", options.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to fetch cluster information: %s", err)
+	}
+	if *clusterInfo.Spec.DatastoreReady {
+		*clusterInfo.Spec.DatastoreReady = false
+		if clusterInfo, err = c.ClusterInformation().Update(ctxt, clusterInfo, options.SetOptions{}); err != nil {
```

Comment:
I think we should wrap this logic in a retry loop - it's likely we'll hit `ErrorResourceUpdateConflict` here. In that case, we can simply retry setting to `false` rather than exit the entire container and start from the beginning.